data_model/src/predicate.rs

```diff
@@ -46,10 +74,83 @@ impl PredicateBox {
             Self::And(preds) => Self::Or(preds.into_iter().map(Self::negate).collect()),
             Self::Or(preds) => Self::And(preds.into_iter().map(Self::negate).collect()),
             Self::Not(pred) => *pred, // TODO: should we recursively simplify?
-            Self::Raw(pred) => Self::Not(Box::new(PredicateBox::Raw(pred))),
+            Self::Raw(pred) => Self::Not(Box::new(Self::Raw(pred))),
+        }
+    }
+}
+
+pub trait PredicateTraitExt<I>: Sized {
```

Comment:
```suggestion
pub trait LogicalCombinator<I>: Sized + core::ops::Neg {
```

Might also want to replace the `negate` with an implementation of `core::ops::Neg` as a trait bound. At the time, given that `&&` and `||` were not overloadable, I thought it made sense to have these in a separate trait. Now it seems that two different implementations of `neg` are redundant.

Reply:
I assume you meant `core::ops::Not`. Implemented.

Reply:
Yes. Typo